test/cpp/tensorexpr/test_cuda.cpp

```diff
@@ -1305,6 +1406,812 @@ void testCudaMaskBlockAndThreadDim() {
+  cudaMalloc(&b_dev, OUTER_SIZE * B_SIZE * sizeof(float));
+  float* c_dev = nullptr;
+  cudaMalloc(&c_dev, OUTER_SIZE * A_SIZE * sizeof(float));
+  float* d_dev = nullptr;
+  cudaMalloc(&d_dev, OUTER_SIZE * B_SIZE * sizeof(float));
+  cudaMemcpy(
+      a_dev,
+      a_v.data(),
+      OUTER_SIZE * A_SIZE * sizeof(float),
+      cudaMemcpyHostToDevice);
+  cudaMemcpy(
+      b_dev,
+      b_v.data(),
+      OUTER_SIZE * B_SIZE * sizeof(float),
+      cudaMemcpyHostToDevice);
+  cudaMemcpy(
+      c_dev,
+      c_v.data(),
+      OUTER_SIZE * A_SIZE * sizeof(float),
+      cudaMemcpyHostToDevice);
+  cudaMemcpy(
+      d_dev,
+      d_v.data(),
+      OUTER_SIZE * B_SIZE * sizeof(float),
+      cudaMemcpyHostToDevice);
+  cudaDeviceSynchronize();
+
+  cuda_cg(a_dev, b_dev, c_dev, d_dev);
+
+  cudaDeviceSynchronize();
+  cudaMemcpy(
+      c_v.data(),
+      c_dev,
+      OUTER_SIZE * A_SIZE * sizeof(float),
+      cudaMemcpyDeviceToHost);
+  cudaMemcpy(
+      d_v.data(),
+      d_dev,
+      OUTER_SIZE * B_SIZE * sizeof(float),
+      cudaMemcpyDeviceToHost);
+  cudaDeviceSynchronize();
+
+  ExpectAllNear(c_v, c_ref, 1e-5);
+  ExpectAllNear(d_v, d_ref, 1e-5);
+
+  cudaFree(a_dev);
+  cudaFree(b_dev);
+  cudaFree(c_dev);
+  cudaFree(d_dev);
+}
+
+// Tests the case with two loops fused into a common parent, which is not bound
+// to any block or thread dimension - however it's two inner loops are bound to
+// the first thread dimenions. This should work just like the MaskThreadDim test
+// where the bigger loop is unmasked but the smaller is masked.
+void testCudaMaskInnerLoopOneBlock() {
+  KernelScope kernel_scope;
+  int OUTER_SIZE = 10;
+  int A_SIZE = 100;
+  int B_SIZE = 50;
+  Buffer a_buf("a", kFloat, {OUTER_SIZE, A_SIZE});
+  Buffer b_buf("b", kFloat, {OUTER_SIZE, B_SIZE});
+  Buffer c_buf("c", kFloat, {OUTER_SIZE, A_SIZE});
+  Buffer d_buf("d", kFloat, {OUTER_SIZE, B_SIZE});
+
+  // Can't build this using Compute and transforms yet.
+  LoopOptions blockBound;
+  blockBound.set_gpu_block_index(0);
+  LoopOptions threadBound;
+  threadBound.set_gpu_thread_index(0);
+  VarHandle i("i", kInt);
+  VarHandle j("j", kInt);
+  VarHandle k("k", kInt);
+
+  Stmt* stmt = For::make(
+      i,
+      0,
+      OUTER_SIZE,
+      Block::make(
+          {For::make(
+               j,
+               0,
+               A_SIZE,
+               Store::make(c_buf, {i, j}, ExprHandle(2) * a_buf(i, j), 1),
+               threadBound),
+           For::make(
+               k,
+               0,
+               B_SIZE,
+               Store::make(d_buf, {i, k}, c_buf(i, k * 2) + b_buf(i, k), 1),
+               threadBound)}));
+
+  stmt = FlattenIndexes(stmt);
+  stmt = IRSimplifier::simplify(stmt);
+
+  CudaCodeGen cuda_cg(stmt, a_buf, b_buf, c_buf, d_buf);
+
+  std::ostringstream oss;
+  oss << *cuda_cg.stmt();
+
+  // The other loop remains the D write is masked.
+  const std::string& verification_pattern =
+      R"IR(
+# CHECK: for (int i = 0; i < 10
```

Comment:
In the commit message, could you list a few generated code examples before and after your changes? So it is easy to see the effect of your change.

Reply:
Examples from the tests you mean?

Reply:
I've updated the comment on this PR.